generators/app/templates/src/services/utils.ts

```diff
@@ -0,0 +1,110 @@
+import { useState, useEffect, useCallback, useRef } from 'react';
+import { ApiErrorResponse, ApiOkResponse, PROBLEM_CODE, ApiResponse } from 'apisauce';
+
+type Error<E> = (error?: { problem: PROBLEM_CODE; errorData?: E }) => void;
+type Request<P, D, E> = (params: P) => Promise<ApiResponse<D, E>>;
+type Success<D> = (data?: D) => void;
+type Failure<E> = (error: { problem: PROBLEM_CODE; errorData?: E }) => void;
+type PostFetch<D, E> = (response: ApiOkResponse<D> | ApiErrorResponse<E>) => void;
+
+interface AsyncRequestHookParams<P, D, E> {
+  request: Request<P, D, E>;
+  payload: P;
+  withPostSuccess?: Success<D>;
+  withPostFailure?: Failure<E>;
+  initialState?: D | null;
+  withPostFetch?: PostFetch<D, E>;
+  shouldExecuteFetch?: boolean;
+  transformResponse?: (response: D | E) => any;
+}
+
+interface AsyncRequest<P, D, E> {
+  values: P;
+  request: Request<P, D, E>;
+  onPrefetch: () => void;
+  onSuccess: Success<D>;
+  onError: Failure<E>;
+  onPostFetch: PostFetch<D, E>;
+}
+
+const executeAsyncRequest = async <P, D, E>({
+  values,
+  request,
+  onPrefetch,
+  onSuccess,
+  onError,
+  onPostFetch
+}: AsyncRequest<P, D, E>) => {
+  onPrefetch();
+  const response = await request(values);
+  if (response.ok) {
+    onSuccess(response.data);
+  } else {
+    onError({ problem: response.problem, errorData: response.data });
+  }
+  onPostFetch(response);
+};
+
+export const useAsyncRequest = <P, D, E>(
+  {
+    request,
+    payload,
+    withPostSuccess,
+    withPostFailure,
+    initialState = null,
+    withPostFetch,
+    shouldExecuteFetch = true,
+    transformResponse = response => response
+  }: AsyncRequestHookParams<P, D, E>,
+  dependencies: []
+) => {
+  const [state, setState] = useState(initialState);
+  const [loading, setLoading] = useState(shouldExecuteFetch || false);
+  const [error, setError] = useState<Error<E> | null>(null);
```

Comment:
why not include the null case type in the interface definition?

Reply:
I'm not sure what you mean. If you want to allow devs to use and endpoint and avoid handling the error, I don't agree. In every endpoint we should have a possible error and handle it. I think this will force us to do that, which I believe it's an improvement. Error should never be null, endpoints can always fail.